### PR TITLE
ENT-5106: Type-hint subscription_manager/ files

### PR DESCRIPTION
**Pull request by m-horky**
* Card ID: ENT-5106

- Along with type hints, some typos in comments have been fixed and some import statements were reordered.

Unfortunately, it is a large PR due to the card scope. Most of the files should be easy to review though, as frequently there were only some minor changes.

Because type hints of function arguments and return values are evaluated at runtime, some have been put into quotes as strings -- they are not being loaded, they are imported only when `TYPE_CHECKING` is True. Variable arguments are not evaluated at runtime, so they do not need to be stringified.

### CODE REVIEW

File: src/subscription_manager/cli.py
Comment (ptoscano) [inline src/subscription_manager/cli.py:183]: hmm this really is `int` -- the problem is that the `try..except` block at the end of it calls `print()` rather than using `sys.exit()`; this "None" return value bubbles up
- ... to `ManagerCLI.main()`
- ... to the `src/subscription_manager/scripts/subscription_manager.py` script in its `main()` function
- ... to the `src/subscription_manager/scripts/subscription_manager.py` script in its `if __name__ == "__main__"` block, which does a sad `main() or 0`...

I think it would be nice to fix/simplify this in another PR

File: src/subscription_manager/managerlib.py
Comment (ptoscano) [inline src/subscription_manager/managerlib.py:121]: this was used only by the GUI code; can you please do a PR before this dropping this dead code? IMHO it'd be better than type-hinting it...

File: src/subscription_manager/managerlib.py
Comment (ptoscano) [inline src/subscription_manager/managerlib.py:355]: extra whitespace changes

File: src/subscription_manager/overrides.py
Comment (ptoscano) [inline src/subscription_manager/overrides.py:79]: i guess `Optional[object]`, just like it is set few lines below for `self.value`?

File: src/subscription_manager/plugins.py
Comment (ptoscano) [inline src/subscription_manager/plugins.py:175]: shouldn't `default` be `Optional[object]`?

File: src/subscription_manager/plugins.py
Comment (ptoscano) [inline src/subscription_manager/plugins.py:520]: i guess `search_path` should be `Optional[str]`?

File: src/subscription_manager/plugins.py
Comment (ptoscano) [inline src/subscription_manager/plugins.py:883]: both should be `Optional[str]`?

File: src/subscription_manager/printing_utils.py
Comment (ptoscano) [inline src/subscription_manager/printing_utils.py:94]: then please create a PR to drop it :)

File: src/subscription_manager/repofile.py
Comment (ptoscano) [inline src/subscription_manager/repofile.py:596]: `Optional[str]` for both, from what I can see

File: src/subscription_manager/repolib.py
Comment (ptoscano) [inline src/subscription_manager/repolib.py:528]: strictly speaking, "advertized" is correct (American English)

File: src/subscription_manager/rhelproduct.py
Comment (ptoscano) [inline src/subscription_manager/rhelproduct.py:39]: maybe `Optional["Product"]`?

File: src/subscription_manager/utils.py
Comment (ptoscano) [inline src/subscription_manager/utils.py:257]: I used the non-integer fallback on purpose: this was the code can know when `get_terminal_size()`  failed and returned the fallback, so the 1000 value can be used

File: src/subscription_manager/utils.py
Comment (ptoscano) [inline src/subscription_manager/utils.py:502]: i guess `Optional[str]` for `filter_string`? also, did you forget `service_level` here? :)

**Review by ptoscano — CHANGES_REQUESTED** [review]
Ooof, lot to go through and check (with associated sadness due to the discovery of horrors done in the past)

I left various notes with things to fix in this PR, and things which would be better to fix outside of this (possibly even before this, simplifying things a bit).

Also, all around both `Dict` and `dict` are used: even if `Dict` is deprecated in Python 3.9, I think it'd be better to use `Dict` all around, so a) there is consistency b) new code that uses `Dict` can be easily backported to 1.28 if needed.

File: src/subscription_manager/listing.py
Comment (jirihnidek) [inline src/subscription_manager/listing.py:13]: There should be some empty line.

File: src/subscription_manager/overrides.py
Comment (jirihnidek) [inline src/subscription_manager/overrides.py:14]: Please add at least empty line between GPL license and first line of code.

File: src/subscription_manager/plugins.py
Comment (jirihnidek) [inline src/subscription_manager/plugins.py:166]: Wouldn't be better to use `Type[SubManPlugin]`? Calling `type()` in type hint looks wild.

File: src/subscription_manager/plugins.py
Comment (jirihnidek) [inline src/subscription_manager/plugins.py:276]: Same comment here.

File: src/subscription_manager/plugins.py
Comment (jirihnidek) [inline src/subscription_manager/plugins.py:311]: And here.

File: src/subscription_manager/plugins.py
Comment (jirihnidek) [inline src/subscription_manager/plugins.py:326]: And here and some other places...

File: src/subscription_manager/syspurposelib.py
Comment (jirihnidek) [inline src/subscription_manager/syspurposelib.py:162]: All these arguments should be `Optional`.

File: src/subscription_manager/utils.py
Comment (jirihnidek) [inline src/subscription_manager/utils.py:184]: The `uep` should be also `Optional["UEPConnection"]`

File: src/subscription_manager/utils.py
Comment (jirihnidek) [inline src/subscription_manager/utils.py:202]: Similar comment here.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for this PR. :+1: I have few small requests. Next time please do not create so big PRs, when it is not necessary. Thanks.

**Review by jirihnidek — APPROVED** [review]
Thanks for updates. :+1: LGTM

File: src/subscription_manager/cert_sorter.py
Comment (ptoscano) [inline src/subscription_manager/cert_sorter.py:68]: `Optional[datetime]`

File: src/subscription_manager/cert_sorter.py
Comment (ptoscano) [inline src/subscription_manager/cert_sorter.py:84]: all of them (except the `bool` one) should be `Optional[...]`

**Review by ptoscano — CHANGES_REQUESTED** [review]
mostly LGTM now; I left a couple of easy notes, and please notice an unfixed note (still about `Optional`)

also please rebase it, so we can get updates and CI improvements

**Review by ptoscano — APPROVED** [review]
Jirka approved, and now I do it too.

Will wait for the successful completion of the CI jobs before merging.

**Comment by ptoscano** [comment]
The `rhel-8-8` job fails, I guess because of incompatibilities with Python < 3.9; would it be possible to fix it?

**Pull request merged**
